client/client.go

```diff
@@ -151,142 +115,221 @@ func New(url string, args ...interface{}) *Client {
+			time.Sleep(500 * time.Millisecond)
+			continue
+		}
 
-		c.connect(url)
-		c.createConsumerChannel()
+		logger.Info("client: finished gracefully")
+		break
 	}
 }
 
-func (c *Client) setupChannels() {
-	messages := c.declareAndConsume()
+// runForever will connect amqp, setup all the amqp channels, run the publisher and run the replies consumer.
+// The method will also return the underlying amqp error if the underlying connection or socket isn't gracefully closed.
+// It will also block until the connection is gone.
+func (c *Client) runOnce(url string) error {
+	logger.Info("client: starting up")
 
-	// queueChannels maps each correlation ID to a channel of *amqp.Delivery.
-	// This is to ensure that no matter the order of a request and response,
-	// we will always publish the response to the correct consumer.
-	// Since we create the correlation ID and hangs waiting for the delivery
-	// channel this is an easy way to ensure no overlapping.
-	// This can of course be a problem if the correlation IDs used is not unique
-	// enough in which case a queueChannel might be overridden if it hasn't
-	// been cleared before.
-	var queueChannels = make(map[string]chan *amqp.Delivery)
+	conn, err := amqp.DialConfig(url, c.dialconfig)
+	if err != nil {
+		return err
+	}
+	defer conn.Close()
+
+	inputCh, err := conn.Channel()
+	if err != nil {
+		return err
+	}
+	defer inputCh.Close()
+
+	outputCh, err := conn.Channel()
+	if err != nil {
+		return err
+	}
+	defer outputCh.Close()
+
+	c.runRepliesConsumer(inputCh)
+	c.runPublisher(outputCh)
+
+	err, ok := <-conn.NotifyClose(make(chan *amqp.Error))
+	if !ok {
+		// The connection was closed gracefully.
+		return nil
+	}
+	// The connection wasn't closed gracefully.
+	return err
+}
 
-	// We spawn a go routine which will listen on the clients message channel.
-	// Each message we receive will be published. If the sender wishes to
-	// receive a response we also map the response channel to the correlation ID.
-	// If we meet an early error before even publishing we send the error back
-	// on that queue.
+// runPublisher consumes messages from clientMessages and publishes them on the amqp exchange.
+// The method will stop consuming if the underlying amqp channel is closed for any reason,
+// and when this happens the messages will be put back in clientMessages unless we have retried to many times.
+func (c *Client) runPublisher(outChan *amqp.Channel) {
+	logger.Info("client: running publisher")
 	go func() {
-		for {
-			request, ok := <-c.clientMessages
-			if !ok {
-				logger.Info("client message channel was closed")
-				break
-			}
+		for request := range c.clientMessages {
+			logger.Infof("client: publishing %v", request.publishing.CorrelationId)
 
 			// Map a request correlation ID to a response channel
-			if request.response != nil {
-				queueChannels[request.publishing.CorrelationId] = request.response
-			}
+			c.mu.Lock()
```

Comment:
Maybe just move this after the `if err != nil` (just before closing the error channel). If so we don't need to use the mutex twice or delete the message if it's beeing requeued.

Reply:
I do it this way because I don't want to keep the lock while I don't need it. While publishing etc. We also don't want to have the lock while we add to any `chan` because they aren't buffered and then we might end up in a deadlock.

Reply:
You don't need to keep the lock since it's no need to lock until we know we will handle the map.

```
diff --git a/client/client.go b/client/client.go
index d0eb0eb..dbdaacc 100644
--- a/client/client.go
+++ b/client/client.go
@@ -215,11 +215,6 @@ func (c *Client) runPublisher(outChan *amqp.Channel) {
                for request := range c.clientMessages {
                        logger.Infof("client: publishing %v", request.publishing.CorrelationId)
 
-                       // Map a request correlation ID to a response channel
-                       c.mu.Lock()
-                       c.correlationMapping[request.publishing.CorrelationId] = request.response
-                       c.mu.Unlock()
-
                        err := outChan.Publish(
                                c.Exchange,
                                request.routingKey,
@@ -229,13 +224,6 @@ func (c *Client) runPublisher(outChan *amqp.Channel) {
                        )
 
                        if err != nil {
-                               // We failed to publish the message so we don't need the correlation mapping any more
-                               // because a new one will be added the next time it's retried. And if it's not retried
-                               // we don't need it any more anyway.
-                               c.mu.Lock()
-                               delete(c.correlationMapping, request.publishing.CorrelationId)
-                               c.mu.Unlock()
-
                                if request.numRetries >= 0 {
                                        // The message that we tried to publish is NOT added back to the
                                        // queue since it never left the client. The sender will get
@@ -254,6 +242,11 @@ func (c *Client) runPublisher(outChan *amqp.Channel) {
                        // The request has left the client, no more errors can occur from here.
                        close(request.errChan)
 
+                       // Map a request correlation ID to a response channel
+                       c.mu.Lock()
+                       c.correlationMapping[request.publishing.CorrelationId] = request.response
+                       c.mu.Unlock()
+
                        logger.Info("client: did publish message")
                }
        }()
```